bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleEngineCallbackImpl.java

```diff
@@ -60,6 +74,21 @@ public boolean isRunning() {
         return future == null || !future.isDone();
     }
 
+    @Override
+    public void addTriggerChangeListener(TriggerChangeListener listener) {
+        if (listeners == null) {
+            listeners = new CopyOnWriteArraySet<TriggerChangeListener>();
```

Comment:
As it is not synchronized this can lead to race conditions where a listener registration can get lost. If you simply instantiate it where it is declared above, then this cannot happen anymore.